web/testing/promise-api/promisehdr.js

```diff
@@ -49,9 +49,13 @@
   function addKeyboardsForLanguage(languages) {
     kmw.addKeyboardsForLanguage(languages.split(',').map(item => item.trim())).then(result => {
       console.log('Adding ' + languages + ': ', result);
-    }).catch(error => {
+    }).catch(errorStubs => {
```

Comment:
This does not receive any error stubs.  It receives Promise rejections; error stubs are included in the array that goes to the `then` clause.

Reply:
The Promise rejections return just error stubs (no successful keyboard stubs)

Reply:
That is not how the `Promise` type definition reads - `Promise<(KeyboardStub|ErrorStub)[]>` basically says that even error stubs should be coming in via the `then` clause.  Also...

https://github.com/keymanapp/keyman/blob/f46d06907a4bd5f9c92cd4d3c0f609d6075967ce/web/source/keyboards/kmwkeyboards.ts#L1236-L1245

Line 1241 supports my position as well - this is what I was referring to when I made that comment.  If this line is going to remain as is, we should be consistent with how `ErrorStub`s are handled + forwarded.

Reply:
I've updated the type definition and converted to try catch
```
let result:(KeyboardStub|ErrorStub)[]|Error = await this.keymanCloudRequest('&keyboardid='+cmd, false);
```

Though when I read the Promise documentation, it sounds like the same type applies to both `resolve` and `reject`.

Reply:
Still isn't handling this scenario well.

A few tests:

- spanish, khmer
    - Returns two stubs without issue
- foo, bar
    - Passes 'error stub' to the `catch` clause, but...
- spanish, foo
    - Passes one keyboard stub + one error stub to the `then` clause.  The page does not present the browser's default alert for the error stub, as the `then` clause doesn't even think to handle it.

Note those last two points; this is inconsistent.  Secondly, the function still completed as suggested.  Yes, we couldn't add keyboards for the languages, but it wasn't an _exceptional_ case - the necessary calls all completed correctly.

What I think should happen:
- foo, bar
    - Passes two error stubs to the `then` clause.  Sure, we didn't find matching keyboards, but we **did** confirm that none are available.  There was no internet issue
- foo, bar (with internet connectivity issues)
    - If the internet connection drops, thus the API call itself fails to complete properly... _**that**_ is a `reject` error.  We couldn't even confirm that foo and bar aren't languages.

Reply:
I think perhaps the reason that we are struggling to come to agreement on how to do this is that we have the wrong promise design in the spec. If instead we return an array of Promises, then it is up to the consumer to decide how they will handle resolve/reject, and we can be clean in our return values for resolve and reject.

For example, you'd use `Promise.all` or `Promise.allSettled`, etc.

https://davidwalsh.name/promises-results gives more info too.

The only question then in my mind is if we return an unwrapped Promise if only a single keyboard is requested.

Reply:
For example,

```ts
let p = keyman.addKeyboardsForLanguage(['en','fr','foo']);
Promise.all(p).then(() => {
 ... // do something with success or failure as you please.
});
```

Reply:
I'm pretty sure it's more a disagreement about when `reject` / `catch` should be used.  I believe that an error stub should not pass through `reject` / `catch`.  Those are for when _actual_ errors prevent the API call from even completing, not for when they complete, but don't have a favorable answer.

"Error stubs" are much closer to "Keyboard stubs" than they are to "internet connection errors".

Reply:
I'm not sure if my idea is possible. The best I can do is to return a Promise which contains an array of Promises, which is unnecessary indirection. So scrap that?

The question is what does the API consumer want: I think delineating on the type of error is not the best API approach. 

If the API user passes in single keyboard, they can use the normal resolve/reject philosophy, assured that *any* rejection means the keyboard hasn't loaded.

If they instead pass in an array of keyboards to load, then they will need to do more work: handle both resolve and reject scenarios, with potential for some errors interspersed through even if resolve is called.

Reply:
> I think delineating on the type of error is not the best API approach.

If an API consumer wants the option to do so, I believe it is important that we cause as little interference as possible.  A lot of programming design is all about providing options when reasonable, so I'd prefer that we leave options open to developers where reasonable.

Throwing an array of error stubs (since conceptually, that's what using `reject` _means_ - hence the `catch` clause) while also throwing literal JS errors through the same parameter is gonna make type checking tricky for delineation of errors.  If they're provided to the 'then' clause, at least the underlying structure of the argument is fairly similar to the "successful stubs" case.

Secondly, because of how `then` and `catch` clause chaining work, a site developer could easily "throw" the error stubs from within the `then` and immediately handle it with an appended `catch` clause, giving effectively the same results seen here.  As opposed to _forcing them_ through the `catch`.

Finally, `catch` is generally meant to handle _exception_al cases - stuff that is straight-up outside of the standard control flow.  Internet errors fit; successful queries that come up negative, less so.  Deviating from the usual paradigm used everywhere for `try`-`catch` stuff, in itself, could give rise to confusion.

Reply:
> If an API consumer wants the option to do so, I believe it is important that we cause as little interference as possible. A lot of programming design is all about providing options when reasonable, so I'd prefer that we leave options open to developers where reasonable.

How are we reducing options? The data is available either way.

> Throwing an array of error stubs (since conceptually, that's what using `reject` _means_ - hence the `catch` clause) while also throwing literal JS errors through the same parameter is gonna make type checking tricky for delineation of errors. If they're provided to the 'then' clause, at least the underlying structure of the argument is fairly similar to the "successful stubs" case.

This doesn't make sense to me. The Promise `reject` function is a _standard_ way of returning failure, regardless of the type of failure. Again, think of the basic case: adding a single keyboard, an API consumer generally _only_ wants to know if it succeeded or failed.

> Finally, `catch` is generally meant to handle _exception_-al cases - stuff that is straight-up outside of the standard control flow. Internet errors fit; successful queries that come up negative, less so. Deviating from the usual paradigm used everywhere for `try`-`catch` stuff, in itself, could give rise to confusion.

Do you have a reference for this? I'm not seeing this kind of delineation anywhere in other APIs that use Promises. This is not about `try`-`catch` -- it's about `Promise.prototype.catch`; the name as used in Promises is perhaps a little misdirected.